### DISCUSSION TITLE
[core][CPD]Skip detecting comments?

### DISCUSSION THREAD

**Discussion by iccengan** [pmd/pmd#3032] [Q&A] [answered]
Since CPD is a duplicate code detection tool, I assume CPD should detect logical lines only. Which means leading comments (for example, method headers) should not be detect, otherwise the duplicate rate of one file can get very high.
Below is one of the test sample. Just want to make sure, does CPD supports skipping leading comment detection?
```
<codefragment><![CDATA[}
/*
asdf
fdsa
asdf
fdsa
asdf
*/
public static void HelloWorld(int i, int j, int k)
{
    int a = i + j;
    int b = j * k;
    int c = i * k;
    int z = 0;
    for(int n = 0; n < 10; n ++)
    {
        z = z + n;
        if(a > 0)
        {
            z += a;
        }
        else if (b > 0)
        {
            z -= b;
        }
        else
        {
            z *= c;
        }
    }
}]]></codefragment>
```

**Comment by iccengan** [comment]
Just found out although code fragment is shown in the detection output, the token number does not count the comment lines. 
One question here, does the detection output the total token number for each file? So that I could calculate the duplicate rate by dividing duplicate token number by total token number.

**Comment by jsotuyod** [comment, marked as answer]
@iccen8 comments are not considered when matching duplicates, but are analyzed to allow for [suppressions](https://pmd.github.io/pmd-6.12.0/pmd_userdocs_cpd.html#suppression)

The total number of tokens per file is not currently exposed to renderers, so you can't currently compute a "duplication percentage".
